src/meta/greedy_load_balancer.cpp

```diff
@@ -903,102 +903,95 @@ void greedy_load_balancer::greedy_balancer(const bool balance_checker)
 void greedy_load_balancer::app_balancer(bool balance_checker)
 {
     const app_mapper &apps = *t_global_view->apps;
-
-    for (const auto &kv : apps) {
-        const std::shared_ptr<app_state> &app = kv.second;
-        if (is_ignored_app(kv.first)) {
-            ddebug_f("skip to do primary balance for the ignored app[{}]", app->get_logname());
-            continue;
-        }
-        if (app->status != app_status::AS_AVAILABLE || app->is_bulk_loading || app->splitting())
-            continue;
-
-        bool enough_information = primary_balancer_per_app(app, _only_move_primary);
-        if (!enough_information) {
-            // Even if we don't have enough info for current app,
-            // the decisions made by previous apps are kept.
-            // t_migration_result->empty();
-            return;
-        }
-        if (!balance_checker) {
-            if (!t_migration_result->empty()) {
-                if (_balancer_in_turn) {
-                    ddebug("stop to handle more apps after we found some actions for %s",
-                           app->get_logname());
-                    return;
-                }
-            }
-        }
+    if (!execute_balance(apps,
+                         balance_checker,
+                         _balancer_in_turn,
+                         _only_move_primary,
+                         std::bind(&greedy_load_balancer::primary_balance,
+                                   this,
+                                   std::placeholders::_1,
+                                   std::placeholders::_2))) {
+        return;
     }
 
-    // TODO: do primary_balancer_globally when we find a good approach to
-    // make decision according to disk load.
-    // primary_balancer_globally();
-
-    if (!balance_checker) {
-        if (!t_migration_result->empty()) {
-            ddebug("stop to do secondary balance coz we already has actions to do");
-            return;
-        }
-    }
-    if (_only_primary_balancer) {
-        ddebug("stop to do secondary balancer coz it is not allowed");
+    if (!need_balance_secondaries(balance_checker)) {
         return;
     }
 
     // we seperate the primary/secondary balancer for 2 reasons:
     // 1. globally primary balancer may make secondary unbalanced
     // 2. in one-by-one mode, a secondary balance decision for an app may be prior than
     // another app's primary balancer if not seperated.
+    execute_balance(apps,
+                    balance_checker,
+                    _balancer_in_turn,
+                    _only_move_primary,
```

Comment:
if `_balancer_in_turn`,` _only_move_primary` are global, so why need pass it? why `_only_primary_balancer` don't like this?

Reply:
cluster load balance will pass value different with global value.

Reply:
why need different? https://github.com/XiaoMi/rdsn/pull/903/files/9b3d0611015a31ae4d9bec75c137e9c71f5cb1e9#diff-ea29ba8c2a8fee984249d3c2431770e76a92a6239971ebb98e21c057017592c5